### ISSUE TITLE
Right way to use AMQPSocketConnection

### ISSUE THREAD

**Issue by goetas** [php-amqplib/php-amqplib#547]
I've been trying to use `AMQPSocketConnection` and setup a consumer that waits for messages.

Actually all the examples provided by the lib are using `AMQPStreamConnection`....

Doing something as:

```php
use PhpAmqpLib\Connection\AMQPSocketConnection;

define('HOST', 'queue');
define('PORT', 5672);
define('USER', 'guest');
define('PASS', 'guest');
define('VHOST', '/');

$exchange = 'basic_get_test';
$queue = 'basic_get_queue';

$connection = new AMQPSocketConnection(HOST, PORT, USER, PASS, VHOST);
$channel = $connection->channel();

$channel->queue_declare($queue, false, true, false, false);
$channel->exchange_declare($exchange, 'fanout', false, true, false);
$channel->queue_bind($queue, $exchange);

$callback = function($msg) {
    echo " [x] Received ", $msg->body, "\n";
};
$channel->basic_consume($queue, '', false, true, false, false, $callback);

while(count($channel->callbacks)) {
    $channel->wait();
}

$channel->close();
$connection->close();
```

This is copied from `amqp_consumer_fanout_1.php` and a bit adapted.

Unfortunately does not work. 

It always throws after 3 seconds of waiting:
```
[16-Feb-2018 15:17:37 UTC] PHP Fatal error:  Uncaught PhpAmqpLib\Exception\AMQPIOException: 
Error reading data. Received 0 instead of expected 7 bytes in /var/www/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/IO/SocketIO.php:131
Stack trace:
#0 /var/www/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/AMQPReader.php(149): PhpAmqpLib\Wire\IO\SocketIO->read(7)
#1 /var/www/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/AMQPReader.php(106): PhpAmqpLib\Wire\AMQPReader->rawread(7)
#2 /var/www/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php(522): PhpAmqpLib\Wire\AMQPReader->read(7)
#3 /var/www/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php(570): PhpAmqpLib\Connection\AbstractConnection->wait_frame(0)
#4 /var/www/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php(225): PhpAmqpLib\Connection\AbstractConnection->wait_channel(1, 0)
#5 /var/www/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php(335): PhpAmqpLib\Channel\AbstractChannel->next_fram in /var/www/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/IO/SocketIO.php on line 131
```

Apparently is because of "read_timeout" set to 3 by default. 

Disabling the read_timeout , "works".
```php
$connection = new AMQPSocketConnection(
   HOST, PORT, USER, PASS, VHOST, 
   false,  // insist
   'AMQPLAIN', // login
   null, // login response
   'en_US', // locale
   0 // <<<-- read timeout
);
```
So this looks working... but, is the right solution?

Will this wait forever? Will it exit in case of errors? will it send heardbeat?

**Comment by creidinger** [comment]
Hi @goetas, Did you ever get a response for this?

**Comment by goetas** [comment]
No really :(

**Comment by lukebakken** [comment]
I've added this to the 2.8.0 milestone and will look at it if I have time. Please remember that this library is maintained by volunteers, so if you're able to look into the issue or provide a fix that is very appreciated.

One good place to look is via a packet capture. I assume that I should be able to reproduce this when I get to it. Thank you for the code.

**Comment by creidinger** [comment]
I don't actually know what changed, but I uninstalled and reinstalled using composer and that resolved my issue.

**Comment by ramunasd** [comment]
Looks like I found issue.
AMQPReader->rawread() should wait for data and only then read it. But when timeout = 0 AMQPReader->wait() does nothing and socket_read() returns empty string without errors. That timeout comes from method AbstractChannel->wait() where it is defined as default argument value.

@goetas for now You can use non-zero timeout in Your loop, something like
```
while(count($channel->callbacks)) {
    try {
        $channel->wait(null, false, 1);
    } catch (\PhpAmqpLib\Exception\AMQPTimeoutException $e) {
    }
}
```

@lukebakken I think default timeout should be set to 1 second or so. Zero timeouts(or no timeouts) also can be supported, socket_select() accepts null as timeout.

**Comment by goetas** [comment]
Will check it!

**Comment by lukebakken** [comment]
Assuming fixed by #642. If not, please follow up.
